_posts/2021-07-13-ci-metrics.md

```diff
@@ -0,0 +1,116 @@
+
+Recently we also added a more abstract [Error Budget](https://grafana-frontdoor.apps.ocp.ci.centos.org/d/budget/cockpit-ci-error-budget) dashboard, which shows the SLIs with a focus on and normalization to the defined SLO error budgets:
+
+![Grafana error budget](/images/grafana-error-budget.png)
+
+## Drawing conclusions
+
+This error budget view now can guide our activity in an objective way. For example, a value of 50% like on "SLI 4: percentage of tests with runtime > 1 h" means that for that particular SLO we used up half of the error budget. The SLO is "95% of all tests take no more than 1 hour to execute", so the error budget is 5%, and the graph tells us that currently roughly 2.5% of tests take longer than one hour. That metric can still become twice as bad before we have to stop the line and go into "infrastructure is on fire" mode.
+
+On the other hand, "SLI 5: percentage of tests with queue time > 5 minutes" is way beyond the budget. The [detailed graph](https://grafana-frontdoor.apps.ocp.ci.centos.org/d/ci/cockpit-ci?orgId=1&viewPanel=6&from=now-7d&to=now) tells us that it's not actually that bad, the average queue time is 5.17 minutes -- just slightly off. We used to have a queue time of a few seconds, so in our case this is an actual bug. We actually know the cause of that -- it will take a few hours to debug, but the details are not important for this post.
+
+But for other cases, it could just be that the queue time has been like that for months, and nobody was bothered by it -- in that case it may be appropriate to redefine the SLO, and bump it to 10 minutes. The initial SLO was defined by gut feelings and polls as well, after all, and they should regularly be reviewed against the team's opinion whether they still accurarely represent "happiness".
+
+## Next Steps
+
+Our current state is a good foundation, but we are not done:
+
+ * There are no alerts yet. Grafana has alerting support, but I have not looked at this at all. We want our infrastructure maintainers to get an email, and/or possibly an IRC bot message.
+ 
+ * Compared to the weather report, the biggest missing feature in our Grafana boards are links to example test runs which violate the SLOs. The [OpenMetrics format](https://prometheus.io/docs/instrumenting/exposition_formats/#openmetrics-text-format) promises support for "exemplars" which fit exactly that use case, so we need to change our export accordingly, and figure out how to represent them in Grafana.
+
+ * We need to define and follow a process how to react to exceeding the error budget. How many/which team members stop working on regular projects to move to infrastructure fixing? Once we fix the problem, it will take some time for the usual "28 days" view to abate and leave the alert zone, even though the problem is fixed; how do we represent that in the default view? How do we communicate this to other teams?
```

Comment:
May I suggest adding this as a second sentence:

"How does the team do a retrospective to determine why the error budget was exceeded?"

Reply:
Done, and reformatted as a sub-enumeration for readability.